File: test_flow.py
import httpx
import psycopg2
import time

BASE_URL = "http://localhost:8080/api"
EMAIL = "xeniassistant@gmail.com"
DB_URL = "postgresql://xeni:xeni_secret@localhost:5432/xeni_db"

print("--- END-TO-END AUTHENTICATION TEST ---")

print("\n1. Registering User...")
# Registration endpoint might be /api/auth/register or just /auth/register depending on gateway router.go.
# We will use /auth/register since the router uses `api.Group("/auth")` without `/api` prefix, wait.
# The gateway is on :8080. If frontend config uses `/api`, the router maps `/api` globally to Echo.
# Let's check router.go: e.Group("/api")? Usually `api.Group("/auth")` is inside an `api` group. Let's assume `/api/auth/register`.
# Let's use `/api/auth/register`. If it fails, I'll switch.
res = httpx.post(
    "http://localhost:8080/api/auth/register",
    json={"name":"Xeni Assistant", "email":EMAIL, "password":"Xeni_2026", "role":"user"},
    timeout=10.0
)
if res.status_code == 404:
    BASE_URL = "http://localhost:8080"
    res = httpx.post(
        "http://localhost:8080/auth/register",
        json={"name":"Xeni Assistant", "email":EMAIL, "password":"Xeni_2026", "role":"user"},
        timeout=10.0
    )

print("Register Response:", res.status_code, res.text)

print("\n2. Fetching OTP from DB...")
time.sleep(1) # wait for DB insert
conn = psycopg2.connect(DB_URL)
cur = conn.cursor()
cur.execute("SELECT otp FROM users WHERE email = %s", (EMAIL,))
row = cur.fetchone()
otp = row[0] if row else None
conn.close()
if not otp:
    print("FAILED TO FETCH OTP! Aborting.")
    exit(1)
print(f"Captured OTP: {otp}")

print("\n3. Verifying Email...")
res = httpx.post(f"{BASE_URL}/auth/verify-email", json={"email":EMAIL, "otp":otp})
print("Verify Response:", res.status_code, res.text)

print("\n4. Testing Login...")
res = httpx.post(f"{BASE_URL}/auth/login", json={"email":EMAIL, "password":"Xeni_2026"})
print("Login Response:", res.status_code, res.text)

print("\n5. Forgot Password...")
res = httpx.post(f"{BASE_URL}/auth/forgot-password", json={"email":EMAIL})
print("Forgot Password Response:", res.status_code, res.text)

print("\n6. Fetching Reset OTP from DB...")
time.sleep(1)
conn = psycopg2.connect(DB_URL)
cur = conn.cursor()
cur.execute("SELECT otp FROM users WHERE email = %s", (EMAIL,))
new_otp = cur.fetchone()[0]
conn.close()
print(f"Captured Reset OTP: {new_otp}")

print("\n7. Reset Password...")
res = httpx.post(f"{BASE_URL}/auth/reset-password", json={"email":EMAIL, "otp":new_otp, "new_password":"Xeni_2027"})
print("Reset Password Response:", res.status_code, res.text)

print("\n8. Testing New Login...")
res = httpx.post(f"{BASE_URL}/auth/login", json={"email":EMAIL, "password":"Xeni_2027"})
print("New Login Response:", res.status_code, res.text)

print("\n--- TEST COMPLETE ---")

File: gateway/test_flow.go
package main

import (
	"bytes"
	"database/sql"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	_ "gorm.io/driver/postgres"
	_ "github.com/lib/pq"
)

var baseURL = "http://localhost:8080/auth"
var email = "xeniassistant@gmail.com"
var dbURL = "postgresql://xeni:xeni_secret@localhost:5432/ai_os_db?sslmode=disable"

func postJSON(url string, payload map[string]interface{}) (int, string) {
	jsonData, _ := json.Marshal(payload)
	req, _ := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	req.Header.Set("Content-Type", "application/json")
	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return 0, err.Error()
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	return resp.StatusCode, string(body)
}

func main() {
	fmt.Println("--- XENI API AUTHENTICATION TEST ---")

	fmt.Println("\n1. Registering User...")
	code, resp := postJSON(baseURL+"/register", map[string]interface{}{
		"name": "Xeni Assistant", "email": email, "password": "Xeni_2026", "role": "user",
	})
	fmt.Printf("Register Response [%d]: %s\n", code, resp)

	fmt.Println("\n--- STOPPING HERE FOR OTP INPUT ---")
	return
	code, resp = postJSON(baseURL+"/verify-email", map[string]interface{}{
		"email": email, "otp": otp.String,
	})
	fmt.Printf("Verify Response [%d]: %s\n", code, resp)

	fmt.Println("\n4. Testing Login...")
	code, resp = postJSON(baseURL+"/login", map[string]interface{}{
		"email": email, "password": "Xeni_2026",
	})
	fmt.Printf("Login Response [%d]: %s\n", code, resp)

	fmt.Println("\n5. Forgot Password...")
	code, resp = postJSON(baseURL+"/forgot-password", map[string]interface{}{
		"email": email,
	})
	fmt.Printf("Forgot Password Response [%d]: %s\n", code, resp)

	fmt.Println("\n6. Fetching Reset OTP from DB...")
	time.Sleep(3 * time.Second)
	var newOtp sql.NullString
	err = db.QueryRow("SELECT otp FROM users WHERE email = $1", email).Scan(&newOtp)
	if err != nil || !newOtp.Valid {
		fmt.Printf("FAILED TO FETCH NEW OTP!\n")
		return
	}
	fmt.Printf("Captured Reset OTP: %s\n", newOtp.String)

	fmt.Println("\n7. Reset Password...")
	code, resp = postJSON(baseURL+"/reset-password", map[string]interface{}{
		"email": email, "otp": newOtp.String, "new_password": "Xeni_2027",
	})
	fmt.Printf("Reset Password Response [%d]: %s\n", code, resp)

	fmt.Println("\n8. Testing New Login...")
	code, resp = postJSON(baseURL+"/login", map[string]interface{}{
		"email": email, "password": "Xeni_2027",
	})
	fmt.Printf("New Login Response [%d]: %s\n", code, resp)

	fmt.Println("\n--- TEST COMPLETE ---")
}
